lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -529,19 +545,82 @@ getLegacyTransaction tid = do
     let inps = mempty
     pure $ Tx tid inps outs
 
+
+putWitness :: TxWitness -> Put
+putWitness (TxWitness bytes) = putByteString bytes
+
 putSignedTx :: [(TxIn, Coin)] -> [TxOut] -> [TxWitness] -> Put
 putSignedTx inputs outputs witnesses = do
-    putTx inputs outputs
+    putInputsOutputs inputs outputs
     unless (length inputs == length witnesses) $
         fail "number of witnesses must equal number of inputs"
     mapM_ putWitness witnesses
-  where
-    -- Assumes the `TxWitness` has been faithfully constructed
-    putWitness :: TxWitness -> Put
-    putWitness (TxWitness bytes) = putByteString bytes
 
-putTx :: [(TxIn, Coin)] -> [TxOut] -> Put
-putTx inputs outputs = do
+-- | Constructs a "normal" tx
+constructTransactionFragment
+    :: Monad m
+    => [(TxIn, TxOut)]
+    -> [TxOut]
+    -> (Address -> m (Hash "sigData" -> TxWitness))
+    -> m (Hash "Tx", SealedTx)
+constructTransactionFragment ins outs mkWit = do
+    constructGenericSignedTx FragmentTransaction
+        mempty
+        ins outs mkWit
+        (const mempty)
+
+-- | Constructs a generic transaction of the form
+--
+-- 1. Payload
+-- 2. Inputs / outputs
+-- 3. Witnesses (including signed hash of 1 and 2)
+-- 4. Payload authentication (calculated using 1,2,3 as input)
+constructGenericSignedTx
+    :: Monad m
+    => FragmentSpec
+    -- ^ The kind of transaction
+    -> Put
+    -- ^ Payload (empty for normal utxo transactions)
+    -> [(TxIn, TxOut)]
+    -> [TxOut]
+    -> (Address -> m (Hash "sigData" -> TxWitness))
+    -- ^ Function for producing witnesses for each address in the inputs.
+    -- E.g by looking up the private key and using it for signing with
+    -- @constructUtxoWitness@
```

Comment:
We'd need a `Passphrase` and some kind of `WitnessUtxo | WitnessLegacyUtxo` too 🤔

Reply:
Hmmm. Passphrase -> yes. `WitnessUTxO` or `WitnessLegacyUtxo` -> no, we are constructing it in this very function!

Reply:
We need to know _which type_  of witness to create though!

Keeping it all in one list is elegant. But I'm not fully convinced.

Also, what bears me against creating 
```haskell
data WitnessType = WitnessUtxo | WitnessLegacyUtxo
```
is that we already have 
```haskell
data TxWitnessTag
    = TxWitnessLegacyUTxO
    | TxWitnessUTxO
    | TxWitnessAccount
    | TxWitnessMultisig
    deriving (Show, Eq)
```
but which we cannot use since we have no encoder for creating account or multisig witnesses yet.